src/services/ansiblePlaybook.ts

```diff
@@ -0,0 +1,179 @@
+import * as child_process from 'child_process';
+import { ExecException } from 'child_process';
+import { URL } from 'url';
+import { promisify } from 'util';
+import {
+  Connection,
+  Diagnostic,
+  DiagnosticSeverity,
+  Position,
+  Range,
+} from 'vscode-languageserver';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { withInterpreter } from '../utils/misc';
+import { WorkspaceFolderContext } from './workspaceManager';
+const exec = promisify(child_process.exec);
+
+/**
+ * Acts as an interface to ansible-playbook command.
+ */
+export class AnsiblePlaybook {
+  private connection: Connection;
+  private context: WorkspaceFolderContext;
+  private useProgressTracker = false;
+
+  constructor(connection: Connection, context: WorkspaceFolderContext) {
+    this.connection = connection;
+    this.context = context;
+    this.useProgressTracker =
+      !!context.clientCapabilities.window?.workDoneProgress;
+  }
+
+  /**
+   * Acts as an interface to ansible-playbook <file> --syntax-check command and a cache of its output.
+   * ansible syntax-check may provide diagnostics for more than just the file for which
+   * it was triggered, and this is reflected in the implementation.
+   *
+   * Perform ansible syntax-check for the given document.
+   */
+  public async doValidate(
+    textDocument: TextDocument
+  ): Promise<Map<string, Diagnostic[]>> {
+    const docPath = new URL(textDocument.uri).pathname;
+    let diagnostics: Map<string, Diagnostic[]> = new Map();
+    let progressTracker;
+    if (this.useProgressTracker) {
+      progressTracker = await this.connection.window.createWorkDoneProgress();
+    }
```

Comment:
It'd be easier if you assigned a fake object here that does nothing.
```suggestion
    const progressTracker = this.useProgressTracker
      ? await this.connection.window.createWorkDoneProgress()
      : SomeDummyObjectWithTheSameInterface();
```
Then, you wouldn't have to care to guard every invocation with an if clause.